fossology/fossology
# feat(rest): oidc based authentication

## Description

Provide OIDC integration for authentication. Users will be redirected to authentication provider and FOSSology can use the tokens to create a login session.

![image](https://user-images.githubusercontent.com/18077542/145162963-4419cf58-6bec-4064-a654-e98c4ef60daa.png)

The login integration uses "Authorization Code" flow of OAuth 2.0 and can be configured using "Customize" page, with ability to read .

![image](https://user-images.githubusercontent.com/18077542/145163193-2de9520c-93c7-47cb-a475-5ea1cf1d2148.png)

The same OAuth provider can be used for REST API authentication as well with "Client Credentials" flow.

![image](https://user-images.githubusercontent.com/18077542/145172972-aa1154a6-43c2-41c9-b6d4-187e65abb95b.png)

The existing token mechanism will still exist and can be configured from `fossology.conf`. A new section `AUTHENTICATION` is added with 2 parameters `provider` and `resttoken`.
Setting `provider` to `external` will prevent logins with username & password. The mechanism for REST authentication can be set by editing `resttoken`.

![image](https://user-images.githubusercontent.com/18077542/145173174-d2889f43-6649-4d74-bee4-e9cb78dc74ad.png)

## How to test

1. Setup OIDC authentication with FOSSology and check if it works.
2. Check if logout mechanism works.
3. Test the REST authentication with old tokens and OIDC.
4. Test if changing `resttoken` in `fossology.conf` effects the token authentication in REST API.
5. Check if changing `provider` in `fossology.conf` to `external` prevents login with username and password.

Closes #515

<a href="https://gitpod.io/#https://github.com/fossology/fossology/pull/2123"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
